### PR TITLE
Allow ThreadStateManager to set "waiting for callstack"

### PR DESCRIPTION
**Pull request by florian-kuebler**
The ThreadStateSlice proto now contains two additional fields, a callstack_id (that does not yet get set at all), and a status, that can be either of:
 - no callstack is and will be attached to this slice
 - a callstack will be attached to this slice; wait for it
 - there is a callstack attached (callstack id contains data)

With this change, we allow ThreadStateManager to actually specify that we are waiting for a callstack to be attached later (in the ProducerEventProcessor).
By now, the manager will not yet get called (which will happen in the next PR).

Test: Unit test
Bug: http://b/235554760

### CODE REVIEW

File: src/GrpcProtos/capture.proto [hunk 1]
```diff
@@ -599,6 +599,28 @@ message ThreadStateSlice {
   // Otherwise, they should be 0 to minimize bandwidth usage.
   uint32 wakeup_tid = 7;
   uint32 wakeup_pid = 8;
+
+  // This is the callstack of the tracepoint that changed the state
+  // of the thread and caused this slice to happen. If the thread
+  // state is runnable then this callstack is coming from wakeup_tid,
```
Comment (dpallotti) [inline src/GrpcProtos/capture.proto:605]: Ok, we need to be careful here as this is messy.

We have `wakeup_reason` that controls whether `wakeup_tid` is valid, but then we have the state itself that controls whether `triggering_callstack_id` is from this thread or `wakeup_tid`. This is explained here but is still awkward when using the proto. 

Would it correct to say that this callstack comes from this thread if `wakeup_reason` is `kNotApplicable`, and from `wakeup_tid` otherwise?

File: src/GrpcProtos/capture.proto [hunk 2]
```diff
@@ -599,6 +599,28 @@ message ThreadStateSlice {
...
+  // otherwise it's coming from the same thread.
+  uint64 triggering_callstack_id = 10;
+
+  // The callstacks for the tracepoints originate from a different event and are
```
Comment (dpallotti) [inline src/GrpcProtos/capture.proto:609]: We should clarify that "event" means ProducerCaptureEvent. The callstack originate from the same tracepoint and the same perf_event_open event, but then we "split" them during processing.

File: src/GrpcProtos/capture.proto [hunk 3]
```diff
@@ -599,6 +599,28 @@ message ThreadStateSlice {
...
+  // not part of the original ProducerCaptureEvent. They will be added by the
+  // ProducerEventProcessor.
+  // If callstack collection is disabled, `kNoCallstack` expresses, that there
```
Comment (dpallotti) [inline src/GrpcProtos/capture.proto:612]:
```suggestion
  // If callstack collection is disabled, `kNoCallstack` expresses that there
```

File: src/GrpcProtos/capture.proto [hunk 4]
```diff
@@ -599,6 +599,28 @@ message ThreadStateSlice {
...
+  // is no callstack to wait for.
+  // On `kWaitingForCallstack`, the ProducerEventProcessor knows, to wait
```
Comment (dpallotti) [inline src/GrpcProtos/capture.proto:614]:
```suggestion
  // On `kWaitingForCallstack`, the ProducerEventProcessor knows to wait with
```

File: src/GrpcProtos/capture.proto [hunk 5]
```diff
@@ -599,6 +599,28 @@ message ThreadStateSlice {
...
+  // sending this event until the callstack has been received.
+  // Once added a callstack, `kCallstackSet` expresses to the client, that there
```
Comment (dpallotti) [inline src/GrpcProtos/capture.proto:616]:
```suggestion
  // Once added a callstack, `kCallstackSet` expresses to the client that there
```

File: src/GrpcProtos/capture.proto (hunk 3 above)
Comment (dpallotti) [reply src/GrpcProtos/capture.proto:612]: Aaah, German commas...

File: src/LinuxTracing/ThreadStateManager.cpp [hunk 6]
```diff
@@ -131,23 +137,29 @@ std::optional<ThreadStateSlice> ThreadStateManager::OnSchedSwitchIn(uint64_t tim
   slice.set_wakeup_reason(open_state.wakeup_reason);
   slice.set_wakeup_tid(open_state.wakeup_tid);
   slice.set_wakeup_pid(open_state.wakeup_pid);
+  if (open_state.wait_for_callstack) {
+    slice.set_callstack_status(orbit_grpc_protos::ThreadStateSlice::kWaitingForCallstack);
```
Comment (dpallotti) [inline src/LinuxTracing/ThreadStateManager.cpp:141]: The comment above says that there are some cases where `open_state.state != ThreadStateSlice::kRunnable`. Would this callstack be from `wakeup_tid` in that case? I assume everything already fits to place?

File: src/GrpcProtos/capture.proto [hunk 7]
```diff
@@ -599,6 +599,28 @@ message ThreadStateSlice {
   // Otherwise, they should be 0 to minimize bandwidth usage.
   uint32 wakeup_tid = 7;
   uint32 wakeup_pid = 8;
+
+  // This is the callstack of the tracepoint that changed the state
+  // of the thread and caused this slice to happen. If the thread
+  // state is runnable then this callstack is coming from wakeup_tid,
+  // otherwise it's coming from the same thread.
+  uint64 triggering_callstack_id = 10;
```
Comment (dpallotti) [inline src/GrpcProtos/capture.proto:607]: I this we should call this `switch_out_or_wakeup_callstack_id`. This is to clarify that there are two clearly separate cases for this callstack, as described above. And `wakeup` makes the association with `wakeup_tid`, `wakeup_reason`, etc.

File: src/LinuxTracing/ThreadStateManager.h [hunk 8]
```diff
@@ -86,6 +91,8 @@ class ThreadStateManager {
     // to the runnable state.
     pid_t wakeup_tid;
     pid_t wakeup_pid;
+
+    bool wait_for_callstack;
```
Comment (dpallotti) [inline src/LinuxTracing/ThreadStateManager.h:95]: I think this should have a comment like the other fields.

File: src/GrpcProtos/capture.proto [hunk 9]
```diff
@@ -599,6 +599,28 @@ message ThreadStateSlice {
...
+  // On `kWaitingForCallstack`, the ProducerEventProcessor knows, to wait
+  // sending this event until the callstack has been received.
+  // Once added a callstack, `kCallstackSet` expresses to the client, that there
+  // is a valid callstack id set.
+  enum CallstackStatus {
+    kNoCallstack = 0;
+    kWaitingForCallstack = 1;
+    kCallstackSet = 2;
+  }
+  CallstackStatus callstack_status = 11;
```
Comment (dpallotti) [inline src/GrpcProtos/capture.proto:623]: The prefix of this should be the same as the prefix of the callstack id field.

Comment (dpallotti) [inline src/GrpcProtos/capture.proto:623]: I think this field should be before the callstack id.

File: src/GrpcProtos/capture.proto [hunk 10]
```diff
@@ -599,6 +599,28 @@ message ThreadStateSlice {
   // Otherwise, they should be 0 to minimize bandwidth usage.
   uint32 wakeup_tid = 7;
   uint32 wakeup_pid = 8;
+
+  // This is the callstack of the tracepoint that changed the state
+  // of the thread and caused this slice to happen. If the thread
```
Comment (dpallotti) [inline src/GrpcProtos/capture.proto:604]: Does the callstack correspond to the beginning of the slice, or to the end of the slice? I assume the beginning? But it's not very clear.

File: src/LinuxTracing/ThreadStateManager.h [hunk 11]
```diff
@@ -52,30 +52,35 @@ class ThreadStateManager {
   void OnNewTask(uint64_t timestamp_ns, pid_t tid, pid_t was_created_by_tid,
                  pid_t was_created_by_pid);
   [[nodiscard]] std::optional<orbit_grpc_protos::ThreadStateSlice> OnSchedWakeup(
-      uint64_t timestamp_ns, pid_t tid, pid_t was_unblocked_by_tid, pid_t was_unblocked_by_pid);
+      uint64_t timestamp_ns, pid_t tid, pid_t was_unblocked_by_tid, pid_t was_unblocked_by_pid,
+      bool wait_for_callstack = false);
```
Comment (dpallotti) [inline src/LinuxTracing/ThreadStateManager.h:56]: I think this should be called `has_wakeup_callstack` or `has_unblocking_callstack` (depending if we want to match the name of the other parameters or the names in the proto, although we should make the names of the other parameters and the names in the proto coincide).

File: src/LinuxTracing/ThreadStateManager.h [hunk 12]
```diff
@@ -52,30 +52,35 @@ class ThreadStateManager {
...
   [[nodiscard]] std::optional<orbit_grpc_protos::ThreadStateSlice> OnSchedSwitchIn(
       uint64_t timestamp_ns, pid_t tid);
   [[nodiscard]] std::optional<orbit_grpc_protos::ThreadStateSlice> OnSchedSwitchOut(
-      uint64_t timestamp_ns, pid_t tid, orbit_grpc_protos::ThreadStateSlice::ThreadState new_state);
+      uint64_t timestamp_ns, pid_t tid, orbit_grpc_protos::ThreadStateSlice::ThreadState new_state,
+      bool wait_for_callstack = false);
```
Comment (dpallotti) [inline src/LinuxTracing/ThreadStateManager.h:61]: And I think that this should be called `has_switch_out_callstack`.

File: src/LinuxTracing/ThreadStateManager.h (hunk 8 above)
Comment (dpallotti) [inline src/LinuxTracing/ThreadStateManager.h:95]: And finally this should probably be called `has_wakeup_or_switch_out_callstack` (same for the constructor parameters).

File: src/LinuxTracing/ThreadStateManagerTest.cpp [hunk 13]
```diff
@@ -38,6 +38,7 @@ TEST(ThreadStateManager, OneThread) {
   EXPECT_EQ(slice->wakeup_reason(), orbit_grpc_protos::ThreadStateSlice::kNotApplicable);
   EXPECT_EQ(slice->wakeup_pid(), kWakeupPidTidWhenWakeupReasonNotApplicable);
   EXPECT_EQ(slice->wakeup_tid(), kWakeupPidTidWhenWakeupReasonNotApplicable);
+  EXPECT_EQ(slice->callstack_status(), orbit_grpc_protos::ThreadStateSlice::kNoCallstack);
```
Comment (dpallotti) [inline src/LinuxTracing/ThreadStateManagerTest.cpp:41]: Even if it doesn't matter, I think we should also check that the callstack id is set to 0, just like `wakeup_pid` and `wakeup_tid` are checked.

File: src/LinuxTracing/ThreadStateManagerTest.cpp [hunk 14]
```diff
@@ -445,6 +471,51 @@ TEST(ThreadStateManager, NoStateChangeWithSwitchIn) {
...
+  slice = manager.OnSchedSwitchIn(400, kTid);
+  EXPECT_EQ(slice->tid(), kTid);
+  EXPECT_EQ(slice->thread_state(), ThreadStateSlice::kRunnable);
+  EXPECT_EQ(slice->duration_ns(), 100);
+  EXPECT_EQ(slice->end_timestamp_ns(), 400);
+  EXPECT_EQ(slice->wakeup_reason(), orbit_grpc_protos::ThreadStateSlice::kUnblocked);
+  EXPECT_EQ(slice->wakeup_pid(), kWasUnblockedByPid);
+  EXPECT_EQ(slice->wakeup_tid(), kWasUnblockedByTid);
+  EXPECT_EQ(slice->callstack_status(), orbit_grpc_protos::ThreadStateSlice::kWaitingForCallstack);
 }
```
Comment (dpallotti) [inline src/LinuxTracing/ThreadStateManagerTest.cpp:519]: Can you add another `OnSchedSwitchOut`, since the previous `OnSchedSwitchOut` was following `OnInitialState`, which is a bit different than following `OnSchedSwitchIn`?

File: src/LinuxTracing/ThreadStateManager.h (hunk 12 above)
Comment (dpallotti) [reply src/LinuxTracing/ThreadStateManager.h:61]: This way we distinguish the meaning of the callstack between `OnSchedWakeup` and `OnSchedSwitchOut`.

**Review by dpallotti — COMMENTED** [review]
With the callstacks having two different origins, the way everything fits together is really hard to follow unless you are really familiar with it. I'm proposing some renamings and clarifications in the hope of making it a tiny bit better.

File: src/GrpcProtos/capture.proto (hunk 10 above)
Comment (florian-kuebler) [reply src/GrpcProtos/capture.proto]: Added a sentence.

File: src/GrpcProtos/capture.proto (hunk 7 above)
Comment (florian-kuebler) [reply src/GrpcProtos/capture.proto]: Done.

File: src/GrpcProtos/capture.proto (hunk 2 above)
Comment (florian-kuebler) [reply src/GrpcProtos/capture.proto]: Rephrased it.

File: src/GrpcProtos/capture.proto (hunk 3 above)
Comment (florian-kuebler) [reply src/GrpcProtos/capture.proto]: I have to admit, this one was horrible :D

File: src/GrpcProtos/capture.proto (hunk 4 above)
Comment (florian-kuebler) [reply src/GrpcProtos/capture.proto]: Done.

File: src/GrpcProtos/capture.proto (hunk 5 above)
Comment (florian-kuebler) [reply src/GrpcProtos/capture.proto]: Done

File: src/GrpcProtos/capture.proto (hunk 9 above)
Comment (florian-kuebler) [reply src/GrpcProtos/capture.proto]: Done.

Comment (florian-kuebler) [reply src/GrpcProtos/capture.proto]: Done.

File: src/LinuxTracing/ThreadStateManager.cpp (hunk 6 above)
Comment (florian-kuebler) [reply src/LinuxTracing/ThreadStateManager.cpp]: My take on this would be, there is no sched_wakeup event happening, so the slice produced here is either a runnable slice (with the callstack coming from wakeup_tid) or the sleeping slice (with the callstack coming from the thread itself).

I don't see why it should not work here.

File: src/LinuxTracing/ThreadStateManager.h (hunk 11 above)
Comment (florian-kuebler) [reply src/LinuxTracing/ThreadStateManager.h]: Done.

File: src/LinuxTracing/ThreadStateManager.h (hunk 12 above)
Comment (florian-kuebler) [reply src/LinuxTracing/ThreadStateManager.h]: Done.

File: src/LinuxTracing/ThreadStateManager.h (hunk 8 above)
Comment (florian-kuebler) [reply src/LinuxTracing/ThreadStateManager.h]: Done.

Comment (florian-kuebler) [reply src/LinuxTracing/ThreadStateManager.h]: Done.

File: src/LinuxTracing/ThreadStateManagerTest.cpp (hunk 13 above)
Comment (florian-kuebler) [reply src/LinuxTracing/ThreadStateManagerTest.cpp]: Done.

File: src/LinuxTracing/ThreadStateManagerTest.cpp (hunk 14 above)
Comment (florian-kuebler) [reply src/LinuxTracing/ThreadStateManagerTest.cpp:593]: Done.

File: src/GrpcProtos/capture.proto [hunk 15]
```diff
@@ -599,6 +599,30 @@ message ThreadStateSlice {
   // Otherwise, they should be 0 to minimize bandwidth usage.
   uint32 wakeup_tid = 7;
   uint32 wakeup_pid = 8;
+
+  // The callstacks for the tracepoints get split during processing and
+  // originate from a different ProducerCaptureEvent. So the callstacks are not
+  // part of the original ProducerCaptureEvent, but will be added by the
```
Comment (dpallotti) [inline src/GrpcProtos/capture.proto:605]:
```suggestion
  // part of the ThreadStateSlice contained in the original ProducerCaptureEvent, but will be added by the
```

File: src/GrpcProtos/capture.proto [hunk 16]
```diff
@@ -599,6 +599,30 @@ message ThreadStateSlice {
...
+    kNoCallstack = 0;
+    kWaitingForCallstack = 1;
+    kCallstackSet = 2;
+  }
+  CallstackStatus switch_out_or_wakeup_callstack_status = 10;
+
+  // This is the callstack of the tracepoint that changed the state
+  // of the thread and caused this slice to happen. The callstack corresponds to
+  // the beginning of the slice.
+  // If the thread state is runnable then this callstack is coming from
```
Comment (dpallotti) [inline src/GrpcProtos/capture.proto:623]: Carrying this over from the previous review round:

We have wakeup_reason that controls whether wakeup_tid is valid, but then we have the state itself that controls whether triggering_callstack_id is from this thread or wakeup_tid. This is explained here but is still awkward when using the proto.

Would it correct to say that this callstack comes from this thread if wakeup_reason is kNotApplicable, and from wakeup_tid otherwise?

File: src/LinuxTracing/ThreadStateManager.h [hunk 17]
```diff
@@ -86,6 +91,11 @@ class ThreadStateManager {
     // to the runnable state.
     pid_t wakeup_tid;
     pid_t wakeup_pid;
+    // We allow the user to collect callstacks on sched_wakeup and sched_switch out events. The next
```
Comment (dpallotti) [inline src/LinuxTracing/ThreadStateManager.h:94]:
```suggestion
    // We allow the user to collect callstacks on sched_wakeup and sched_switch out events. This
```

File: src/LinuxTracing/ThreadStateManager.cpp (hunk 6 above)
Comment (dpallotti) [reply src/LinuxTracing/ThreadStateManager.cpp]: Makes sense. I was wondering if I had gotten everything correctly.

**Review by dpallotti — COMMENTED** [review]
I missed what I think is an important comment from the previous reviews, I wrote it again.

File: src/GrpcProtos/capture.proto (hunk 1 above)
Comment (florian-kuebler) [reply src/GrpcProtos/capture.proto]: Yeah, I think this is true.

File: src/GrpcProtos/capture.proto (hunk 16 above)
Comment (florian-kuebler) [reply src/GrpcProtos/capture.proto:623]: For this round of reviews. Yep, I think it is correct.

File: src/GrpcProtos/capture.proto [hunk 18]
```diff
@@ -599,6 +599,30 @@ message ThreadStateSlice {
   // Otherwise, they should be 0 to minimize bandwidth usage.
   uint32 wakeup_tid = 7;
   uint32 wakeup_pid = 8;
+
+  // The callstacks for the tracepoints get split during processing and
+  // originate from a different ProducerCaptureEvent. So the callstacks are not
+  // part of the ThreadStateSce contained in the original ProducerCaptureEvent,
```
Comment (dpallotti) [inline src/GrpcProtos/capture.proto:605]:
```suggestion
  // part of the ThreadStateSlice contained in the original ProducerCaptureEvent,
```

File: src/GrpcProtos/capture.proto (hunk 16 above)
Comment (dpallotti) [reply src/GrpcProtos/capture.proto:623]: Then I think we should write it in those terms for the reasons I mentioned: it's better to reason entirely in terms of the `wakeup_` fields rather than having to involve the specific state (and keep track which state implies one thing and which implies the other).

Comment (florian-kuebler) [reply src/GrpcProtos/capture.proto]: Agree. Better now?

File: src/GrpcProtos/capture.proto [hunk 19]
```diff
@@ -599,6 +599,29 @@ message ThreadStateSlice {
...
+    kWaitingForCallstack = 1;
+    kCallstackSet = 2;
+  }
+  CallstackStatus switch_out_or_wakeup_callstack_status = 10;
+
+  // This is the callstack of the tracepoint that changed the state
+  // of the thread and caused this slice to happen. The callstack corresponds to
+  // the beginning of the slice.
+  // If the wakeup_reason is kNotApplicable then this callstack is coming from
+  // the same thread, otherwise it's coming from wakeup_tid.
```
Comment (dpallotti) [inline src/GrpcProtos/capture.proto:623]:
```suggestion
  // the thread this slice belongs to, otherwise it's coming from wakeup_tid.
```
For extra extra clarity.

File: src/GrpcProtos/capture.proto (hunk 16 above)
Comment (dpallotti) [reply src/GrpcProtos/capture.proto]: Yep, thanks.

**Review by dpallotti — APPROVED** [review]
Thanks!

File: src/GrpcProtos/capture.proto [hunk 20]
```diff
@@ -599,6 +599,29 @@ message ThreadStateSlice {
...
+
+  // The callstacks for the tracepoints get split during processing and
+  // originate from a different ProducerCaptureEvent. So the callstacks are not
+  // part of the ThreadStateSlice contained in the original
+  // ProducerCaptureEvent, but will be added by the ProducerEventProcessor. If
+  // callstack collection is disabled, `kNoCallstack` expresses that there is no
+  // callstack to wait for. On `kWaitingForCallstack`, the
+  // ProducerEventProcessor knows to wait with sending this event until the
+  // callstack has been received. Once added a callstack, `kCallstackSet`
+  // expresses to the client that there is a valid callstack id set.
```
[the remark this answers is not in the document]
Comment (florian-kuebler) [reply src/GrpcProtos/capture.proto]: Done.

File: src/GrpcProtos/capture.proto (hunk 19 above)
Comment (florian-kuebler) [reply src/GrpcProtos/capture.proto]: Done.

File: src/LinuxTracing/ThreadStateManager.h [hunk 21]
```diff
@@ -52,30 +52,35 @@ class ThreadStateManager {
...
+      uint64_t timestamp_ns, pid_t tid, orbit_grpc_protos::ThreadStateSlice::ThreadState new_state,
+      bool has_switch_out_callstack = false);
   [[nodiscard]] std::vector<orbit_grpc_protos::ThreadStateSlice> OnCaptureFinished(
       uint64_t timestamp_ns);
 
  private:
   struct OpenState {
-    OpenState(orbit_grpc_protos::ThreadStateSlice::ThreadState state, uint64_t begin_timestamp_ns)
+    OpenState(orbit_grpc_protos::ThreadStateSlice::ThreadState state, uint64_t begin_timestamp_ns,
+              bool wait_for_callstack = false)
```
Comment (dpallotti) [inline src/LinuxTracing/ThreadStateManager.h:68]: @florian-kuebler I found this leftover `wait_for_callstack`; could you rename it as part of https://github.com/google/orbit/pull/4180?

Comment (florian-kuebler) [reply src/LinuxTracing/ThreadStateManager.h:68]: Done.
